Issue #1423: IEJoin Implementation

hawkfish opened the pull request:
Add [IEJoin](https://vldb.org/pvldb/vol8/p2074-khayyat.pdf) operator for faster performance of double range predicates of the form:

```sql
SELECT east.id, west.t id FROM east, west
WHERE east.dur < west.time AND east.rev > west.cost;
```
Benchmarking shows that the performance is 1-2 orders of magnitude faster than our existing operators (NestedLoopJoin and PiecewiseMergeJoin) for this type of query, especially when considering that practical use cases should not significantly increase the result set cardinality.

Some implementation notes:
* Only the "Union Arrays on Join Attributes" version from §4.2 was implemented, as the other variants are more complex and less space efficient;
* The paper describes incorrect handling of partial equality predicates, especially sort order and backtracking;
* The Bloom filter size has been empirically determined to be optimal at 1024:1 on an 18 core iMac Pro;
* The operator is a double pipeline breaker and the side being built in the `Sink` phases is tracked in the global state;
* Matching pairs and unmatched outer entries are emitted in the `Source` phase;
* The distributed version (Algorithm 3) is implemented by joining blocks on parallel threads;
* Additional predicates and outer joins are supported;
* "Simple" joins are not supported and are deferred to NLJ (this is future work);

Commit:
Issue #1423: IEJoin First test

First correctness test west vs west.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: IEJoin algorithm validation

Various fixes to get IEJoin to pass all tests
that involve exactly two range queries.

Commit:
Issue #1423: IEJoin Bloom-ish filter

Check entry blocks of B in lieu of a real Bloom filter.
Improves Employees 10K benchmark (included) by 2x.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: IEJoin algorithm validation

Replace "and" with &&

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Merge branch 'hawkfish-iejoin' of github.com:hawkfish/duckdb into hawkfish-iejoin

Commit:
Issue #1423: IEJoin Left Gather

Switch to left side gathering from simple slicing
in anticipation of larger blocks.

Commit:
Issue #1423: IEJoin Pipelining

Convert IEJoin to a double pipeline breaker.
This requires:
* Building the two sides sequentially using Sink
* Tracking the child being built in the global sink state
* Always being a source and generating all data there.

Commit:
Issue #1423: IEJoin Multiple predicates

Hook up and test extra predicates beyond the required two range predicates.

Commit:
Issue #1423: IEJoin Employees Benchmark

Add description from Khayyat et al.

Commit:
Issue #1423: IEJoin op1 filtering

Fast exit from blocks with no overlap
for X op1 X'

Commit:
Issue #1423: IEJoin Parallelism

Parallel source and logic fixes.

Commit:
Issue #1423: IEJoin Bloom Filters

Add a 1024:1 Bloom filter to accelerate bitmap scanning.
The value was determined empirically
and differs slightly from the paper.

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Issue #1423: IEJoin Block Pairs

Replace list of pairs with modular arithmetic.

Commit:
Issue #1423: IEJoin Outer Performance

Return dense blocks of outer join values.

Commit:
Issue #1423: IEJoin Block Indexing

Fix reversed left/right block index generation.

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+		case ExpressionType::COMPARE_NOTEQUAL:
+		case ExpressionType::COMPARE_DISTINCT_FROM:
+			// Allowed in multi-predicate joins, but can't be first/sort.
+			conditions[--other_position] = std::move(conditions_p[i]);
```
This appears to be untested

hawkfish replied:
Ah, there is a benchmark that does this, but I'll add a small version of Q2 (I still don't have a decent generator for their Events table...)

hawkfish replied:
Heh - it has to be a small test because the unoptimised query melts down at 30KR(!)

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+
+			local_sort_state.SinkChunk(join_head, input);
+		} else {
+			local_sort_state.SinkChunk(keys, input);
```
This appears to be untested

hawkfish replied:
In the back of my mind is a later unification of the utilities in IEJoin, PWMJ and maybe a planar sweep algorithm, so the merged code would use this branch, but I will remove it for now.

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+
+	// Start the iejoin phase or finish if a iejoin is not necessary
+	if (global_sort_state.sorted_blocks.size() > 1) {
+		PhysicalIEJoin::ScheduleMergeTasks(pipeline, event, table);
```
This does not appear to be tested - could we add a test with a bigger IE join?

hawkfish replied:
I can, but I will have to turn off validation.

hawkfish replied:
I'd also like to make this "parallel sort child input" operation shared, so we can get coverage once instead of repeating it for each physical operator.

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+		if (all_constant) {
+			comp_res = FastMemcmp(entry_ptr, other.entry_ptr, cmp_size);
+		} else {
+			comp_res = Comparators::CompareTuple(scan, other.scan, entry_ptr, other.entry_ptr, sort_layout, external);
```
This appears to be untested as well - could we add a test with IE joins with arrays/strings?

hawkfish replied:
Good call - there was a missing `PinData` call in the iterators.

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+					state.right_executor.ExecuteExpression(cmp_idx, right);
+
+					if (tail_count < result_count) {
+						left.Slice(*sel, tail_count);
```
This appears to be untested - could we add a test that triggers this?

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+
+void PhysicalJoin::ConstructRightJoinResult(DataChunk &right, DataChunk &result, bool found_match[]) {
+	SelectionVector remaining_sel(STANDARD_VECTOR_SIZE);
+	idx_t remaining_count = 0;
```
This function appears to be unused

hawkfish replied:
Yes, it was left over from a previous (slow!) right join implementation.  Should I remove it or move it next to the corresponding left join constructor?

Mytherin commented on test/sql/join/iejoin/test_iejoin_overlaps.test:
```diff
@@ -0,0 +1,21 @@
+statement ok
+PRAGMA enable_verification
+
+# We read from CSVs to prevent the optimiser from
```
Can we get some more tests for the IEJoin?

* IEJoin in subqueries
* IEJoin in recursive ctes
* Larger IE Joins and IE joins with arrays/strings as indicated above

Mytherin commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+			return;
+		} else {
+			// found matches: extract them
+			chunk.Reset();
```
nit: this else can go to reduce indentation depth here

hawkfish replied:
Ah, there used to be a body in the return - sorry!

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some requests for more tests and minor nitpicks:

lnkuiper commented on src/execution/operator/join/physical_iejoin.cpp:
```diff
@@ -0,0 +1,1415 @@
+		return *this;
+	}
+
+	inline bool Compare(const SBIterator &other) const {
```
Looks great, good idea to create an iterator. I was wondering if it's a good idea to define more operators so the code becomes less verbose? e.g.: `operator!=`

hawkfish commented on src/common/sort/sorted_block.cpp:
```diff
@@ -281,21 +281,35 @@ void SBScanState::SetIndices(idx_t block_idx_to, idx_t entry_idx_to) {
 	auto count = MinValue((idx_t)STANDARD_VECTOR_SIZE, total_count - total_scanned);
 	if (count == 0) {
-		D_ASSERT(read_state.block_idx == sorted_data.data_blocks.size());
+		// D_ASSERT(read_state.block_idx == sorted_data.data_blocks.size());
```
Yes, it is no longer a valid assertion for single block scanning. I will remove it.

Commit:
Issue #1423: IEJoin String Compare

* Fix string compare issues;
* Disable RTE planning;
* Add more test coverage;

Commit:
Issue #1423: IEJoin Test Stability

* Don't run large joins in 32 bit
* Don't generate random data with vector size smaller than the data

hawkfish committed:
Merge pull request #69 from duckdb/master

Issue #1423: IEJoin Python merge

Commit:
Merge from master.

Commit:
Merge branch 'hawkfish-iejoin' of github.com:hawkfish/duckdb into hawkfish-iejoin

Commit:
CI debugging - what query is hanging?

Commit:
Issue #1423: IEJoin Outer Yield

Toss in a yield() call to see if it fixes the outer join hang.

Commit:
Issue #1423: IEJoin Outer Yield

Yield when waiting to start outer join pair generation.
This seems to prevent hangs .

Commit:
Merge branch 'master' into hawkfish-iejoin

Commit:
Merge branch 'hawkfish-iejoin' into hawkfish-iejoin-hang

Commit:
Issue #1423: IEJoin Unitialised Variable

Initialise atomic counter `completed` to 0 to prevent infinite hangs.

Mytherin commented:
Thanks for the updates!
